### PR TITLE
Add structure to the code

### PR DESCRIPTION
**Pull request by arastoob**
This will help us have a structural code. The main boxes are organized in their own structures and later we can add more functionalities to them, e.g., add some functions to the `Movie` structure to get the corresponding video frame of a specific time period.

### CODE REVIEW

File: src/header.rs [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+#[derive(Debug, Default)]
+pub struct Header {
+    pub name: String,
+    pub size: usize,
+    pub start: usize,
+}
+
+impl Header {
+    pub fn header(data: &[u8], start: usize) -> Result<Self, Error> {
```
Comment (sdharar) [inline src/header.rs]: warning: constructor `header` has the same name as the type
  --> src\header.rs:17:5
   |
17 | /     pub fn header(data: &[u8], start: usize) -> Result<Self, Error> {
18 | |         // the first 8 bytes includes the atom size and its name
19 | |         // The size is the entire size of the box, including the size and type header, fields, and all contained boxes.
20 | |         let size = BigEndian::read_u32(&data[0..4]) as usize;
...  |
23 | |         Ok(Self { name, size, start })
24 | |     }
   | |_____^
   |
   = note: `#[warn(clippy::self_named_constructors)]` on by default
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#self_named_constructors

File: src/boxes/file_type.rs [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+ */
+
+use crate::boxes::Mp4Box;
+use crate::Error;
+use crate::Header;
+
+#[derive(Debug)]
+pub struct FileType {
+    header: Header,
+    data: Vec<u8>,
```
Comment (sdharar) [inline src/boxes/file_type.rs:13]: warning: field is never read: `data`
  --> src\boxes\file_type.rs:13:5
   |
13 |     data: Vec<u8>,
   |     ^^^^^^^^^^^^^
   |
   = note: `#[warn(dead_code)]` on by default

File: src/boxes/media_data.rs [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+ */
+
+use crate::boxes::Mp4Box;
+use crate::Error;
+use crate::Header;
+
+#[derive(Debug)]
+pub struct MediaData {
+    header: Header,
+    data: Vec<u8>,
```
Comment (sdharar) [inline src/boxes/media_data.rs:13]: warning: field is never read: `data`
  --> src\boxes\media_data.rs:13:5
   |
13 |     data: Vec<u8>,

File: src/boxes/mp4_box.rs [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+    fn size(&self) -> usize;
+
+    /// The box name
+    fn name(&self) -> &str;
+
+    /// Read the box content
+    fn read(&self) -> Result<Vec<u8>, Error>;
+
+    /// Get the internal boxes of this box
+    fn fields(&self) -> Option<Vec<&Box<dyn Mp4Box>>>;
```
Comment (sdharar) [inline src/boxes/mp4_box.rs]: warning: you seem to be trying to use `&Box<T>`. Consider using just `&T`
  --> src\boxes\mp4_box.rs:31:36
   |
31 |     fn fields(&self) -> Option<Vec<&Box<dyn Mp4Box>>>;
   |                                    ^^^^^^^^^^^^^^^^ help: try: `&dyn Mp4Box`
   |
   = note: `#[warn(clippy::borrowed_box)]` on by default
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#borrowed_box

File: src/boxes/mp4_box.rs [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+        write!(
+            f,
+            "{}: {{start address: {}, size: {}}} \n",
+            self.name(),
+            self.start(),
+            self.size()
+        )?;
+
+        let internals = self.fields();
+        if internals.is_some() {
```
Comment (sdharar) [inline src/boxes/mp4_box.rs]: warning: called `unwrap` on `internals` after checking its variant with `is_some`
  --> src\boxes\mp4_box.rs:51:29
   |
50 |         if internals.is_some() {
   |         ---------------------- help: try: `if let Some(..) = internals`
51 |             for internal in internals.unwrap() {
   |                             ^^^^^^^^^^^^^^^^^^
   |
   = note: `#[warn(clippy::unnecessary_unwrap)]` on by default
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#unnecessary_unwrap

File: src/bin/main.rs [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+use structopt::StructOpt;
+
+#[derive(Debug, StructOpt)]
+#[structopt(
+    name = "fmv-parser",
+    about = "Parse fmv files and extract the internal structures."
+)]
+struct Opt {
+    #[structopt(short, long)]
+    verbose: bool,
```
Comment (sdharar) [inline src/bin/main.rs]: warning: field is never read: `verbose`
  --> src\bin\main.rs:16:5
   |
16 |     verbose: bool,
   |     ^^^^^^^^^^^^^
   |
   = note: `#[warn(dead_code)]` on by default

**Review by sdharar — CHANGES_REQUESTED** [review]
Dr. @arastoob => The structure of the code has improved a lot, and it appears to generate the expected results. Upon running some tests, I have stumbled upon some changes that have been added as comments.

If you think the code requires changes based on my comments, perhaps those changes can be made before committing this PR.

**Comment by arastoob** [comment]
> 
@sdharar  What comments do you mean? If you mean `reader` and `boxes_old` files, they are your file and I thought we might need it. If you think that the new code has covered what is inside those files, we can remove them?

**Comment by sdharar** [comment]
> 
No, I meant the comments that I added on your code. Yes, the new code does pretty much outputs the same thing that older code did, in addition to one more layer further down in track/moov atoms but its more structured. The ftype atom needs more further layer to be implemented. Perhaps we can take some implementations from the older code for ftype atom. What do you think?

**Comment by arastoob** [comment]
> > 
> 
> No, I meant the comments that I added on your code. Yes, the new code does pretty much outputs the same thing that older code did, in addition to one more layer further down in track/moov atoms but its more structured. The ftype atom needs more further layer to be implemented. Perhaps we can take some implementations from the older code for ftype atom. What do you think?

So where can I see the comments?
and yes, please go ahead and add your code to ftype.

**Comment by sdharar** [comment]
> > > 
> > 
> > 
> > No, I meant the comments that I added on your code. Yes, the new code does pretty much outputs the same thing that older code did, in addition to one more layer further down in track/moov atoms but its more structured. The ftype atom needs more further layer to be implemented. Perhaps we can take some implementations from the older code for ftype atom. What do you think?
> 
> So where can I see the comments? and yes, please go ahead and add your code to ftype.

I added suggestions I received from clippy on this webpage, which I guess you already have rectified.

**Comment by arastoob** [comment]
> > > > 
> > > 
> > > 
> > > No, I meant the comments that I added on your code. Yes, the new code does pretty much outputs the same thing that older code did, in addition to one more layer further down in track/moov atoms but its more structured. The ftype atom needs more further layer to be implemented. Perhaps we can take some implementations from the older code for ftype atom. What do you think?
> > 
> > 
> > So where can I see the comments? and yes, please go ahead and add your code to ftype.
> 
> I added suggestions I received from clippy on this webpage, which I guess you already have rectified.

Yes, I have resolved those comments. Sorry, I thought that you have made some comments inside the code. So, I can merge the code into master.

**Pull request merged**